src/Shared/TryParseMethodCache.cs

```diff
@@ -81,5 +180,110 @@ private static MethodInfo GetEnumTryParseMethod()
             Debug.Fail("static bool System.Enum.TryParse<TEnum>(string? value, out TEnum result) not found.");
             throw new Exception("static bool System.Enum.TryParse<TEnum>(string? value, out TEnum result) not found.");
         }
+
+        private static bool TryGetDateTimeTryPareMethod(Type type, out MethodInfo? methodInfo)
+        {
+            methodInfo = null;
+            if (type != typeof(DateTime) && type != typeof(DateOnly) &&
+                 type != typeof(DateTimeOffset) && type != typeof(TimeOnly))
+            {
+                return false;
+            }
+
+            var staticDateMethods = type.GetMethods(BindingFlags.Public | BindingFlags.Static)
+                   .Where(m => m.Name == "TryParse" && m.ReturnType == typeof(bool))
+                   .OrderByDescending(m => m.GetParameters().Length);
+
+            foreach (var method in staticDateMethods)
+            {
+                var tryParseParameters = method.GetParameters();
+
+                if (tryParseParameters.Length == 4 &&
+                    tryParseParameters[0].ParameterType == typeof(string) &&
+                    tryParseParameters[1].ParameterType == typeof(IFormatProvider) &&
+                    tryParseParameters[2].ParameterType == typeof(DateTimeStyles) &&
+                    tryParseParameters[3].IsOut &&
+                    tryParseParameters[3].ParameterType == type.MakeByRefType())
+                {
+                    methodInfo = method;
+                    break;
+                }
+            }
+
+            return methodInfo != null;
+        }
+
+        private static bool TryGetNumberStylesTryGetMethod(Type type, out MethodInfo? method, out NumberStyles? numberStyles)
```

Comment:
```suggestion
        private static bool TryGetNumberStylesTryGetMethod(Type type, [NotNullWhen(true)] out MethodInfo? method, [NotNullWhen(true)] out NumberStyles? numberStyles)
```

This should save some `!`.